pretyping/constr_matching.ml

```diff
@@ -409,9 +409,9 @@ let matches_head env sigma pat c =
   matches env sigma pat head
 
 (* Tells if it is an authorized occurrence and if the instance is closed *)
-let authorized_occ env sigma partial_app closed pat c mk_ctx =
+let authorized_occ env sigma closed pat c mk_ctx =
   try
-    let subst = matches_core_closed env sigma false partial_app pat c in
+    let subst = matches_core_closed env sigma false true pat c in
```

Comment:
With this it seems `matches_core_closed` is always called with `(false,true)` or `(true,false)`. Is there a reason for that or is it just coincidence?

Reply:
Yeah, this is hardwiring some semantics facts about Ltac, e.g. the fact that we want to collect bound names as well. We're never using the other combinations, because either we want a sane behaviour or an *expressive* one.

Reply:
Note though that the only call of `matches_core_closed` with `(true, false)` is in the definition of `is_matching_conv` and the only use of the latter is in the definition of `pf_is_matching` and the only use of the latter is in `plugins/funind/recdef.ml`...

Reply:
@Zimmi48 No, it's also used in Hipattern by `match_eq_nf`.

Reply:
Apart from the one use in Funind, a full search on all the code on GitHub shows that this is also used twice in https://github.com/coq-contribs/rational/blob/master/Rewrite/GenericRewrite/HeadSimpl/tacentry.ml4. That's it.

Reply:
@ppedrot You're right and that's because I missed that `matches_conv` was also exported.

Reply:
@Zimmi48 That said, in both cases the code can be rewritten in a straightforward way that does not rely on syntactic matching. I don't quite know why it was written like that in the first place.

(Well, for `funind` I definitely have my own opinion.)

Reply:
About the `rational` contribution: it was the first version of ring by Samuel Boutin, before Patrick Loiseleur made the next version named 'ring', before Assia, Bruno and Benjamin made the current version of ring.

I stopped porting it as it is one of those which deserves the name "museum".

For the anecdote, this version was called `Polynom` and this is why the chapter in the manual has still the name `Polynom.tex`.

Reply:
@Zimmi48 I'll submit a PR that removes the `matches_conv` stuff.

Reply:
@herbelin Interesting bits of history! I think it would make sense to actually create a "museum" (a page on the Coq website, with links to some old code and historical notes). At the same time, the contribs that are still useful (I'm thinking about aac-tactic in particular) could be moved from the @coq-contribs GitHub organization to the @coq organization (and get the same visibility as bignums, and possibly be shipped in a "Coq distribution" see #6208).

Reply:
> I think it would make sense to actually create a "museum" (a page on the Coq website, with links to some old code and historical notes).

I'd would like that too.

> At the same time, the contribs that are still useful (I'm thinking about aac-tactic in particular) could be moved from the @coq-contribs GitHub organization to the @coq organization (and get the same visibility as bignums, and possibly be shipped in a "Coq distribution" see #6208).

+1

Note: Actually, my memory is fuzzy. Maybe `Polynom` was the initial name of Loiseleur's version and it was later released as `Ring`. Someone in Orsay at this time would know better.